src/actions/plugins/imswitcher.ts

```diff
@@ -0,0 +1,144 @@
+import { ModeName } from '../../mode/mode';
+import { existsSync } from 'fs';
+import { configuration } from '../../configuration/configuration';
+import * as util from '../../util/util';
+import { logger } from '../../util/logger';
+import { Message } from '../../util/message';
+import { Globals } from '../../globals';
+
+// InputMethodSwitcher change input method automatically when mode changed
+export class InputMethodSwitcher {
+  constructor(execute: (cmd: string) => Promise<string> = util.executeShell) {
+    this.execute = execute;
+  }
+
+  private savedIMKey = '';
+  private execute: (cmd: string) => Promise<string>;
+
+  public async switchInputMethod(prevMode: ModeName, newMode: ModeName) {
+    if (configuration.autoSwitchInputMethod.enable !== true) {
+      return;
+    }
+    if (!this.validateConfiguration()) {
+      this.disableIMSwitch();
+      return;
+    }
+    // when you exit from insert-like mode, save origin input method and set it to default
+    if (
+      prevMode === ModeName.Insert ||
```

Comment:
Can you put this put `ModeName.Insert, ModeName.SurroundInsertMode, ModeName.Replace` into a hashset? Then this check can be simplified to...

```
if (newMode !== prevMode) {
   if (insertLikeMode.has(newMode)) {
   // do whatever
   } else if (insertLikeMode.has(oldMode)) {
   // do whatever
   }
}
```

Is that right? Hopefully my logic is correct there.

Reply:
What if prevMode is `Insert` and newMode is `SurroundInputMode`?
In your logic:
```
if (newMode !== prevMode) {
   if (insertLikeMode.has(newMode)) {
      //It will hit this spot
   } else if (insertLikeMode.has(oldMode)) {

   }
}
```
In fact, It should do noting in this situation!

Reply:
I use a new concept: ModeCluster to simplify this check, How do you like it?